### ISSUE TITLE
Invalid container image name

### ISSUE THREAD

**Issue by AFCMS** [actions/attest-build-provenance#71]
I have tried attesting the Docker images built by my GitHub Actions workflow and pushed to ghcr.io.

I have tried using the documentation [here](https://docs.github.com/en/actions/publishing-packages/publishing-docker-images) and [here](https://github.com/actions/attest-build-provenance?tab=readme-ov-file#container-image).

I have the following workflow steps ([source here](https://github.com/AFCMS/minetest-skin-server/blob/master/.github/workflows/build-image.yml)):

```yml
env:
  # Use docker.io for Docker Hub if empty
  REGISTRY: ghcr.io
  # github.repository as <account>/<repo>
  IMAGE_NAME: ${{ github.repository }}
....
      # Login against a Docker registry except on PR
      # https://github.com/docker/login-action
      - name: Log into registry ${{ env.REGISTRY }}
        if: github.event_name != 'pull_request'
        uses: docker/login-action@v3.1.0
        with:
          registry: ${{ env.REGISTRY }}
          username: ${{ github.actor }}
          password: ${{ secrets.GITHUB_TOKEN }}

      # Extract metadata (tags, labels) for Docker
      # https://github.com/docker/metadata-action
      - name: Extract Docker metadata
        id: meta
        uses: docker/metadata-action@v5.5.1
        with:
          images: ${{ env.REGISTRY }}/${{ env.IMAGE_NAME }}
          labels: |
            org.opencontainers.image.title=Minetest Skin Server

      # Build and push Docker image with Buildx (don't push on PR)
      # https://github.com/docker/build-push-action
      - name: Build and push Docker image
        id: build-and-push
        uses: docker/build-push-action@v5.3.0
        with:
          context: .
          platforms: linux/amd64,linux/arm64,linux/386,linux/ppc64le,linux/s390x
          push: ${{ github.event_name != 'pull_request' }}
          tags: ${{ steps.meta.outputs.tags }}
          labels: ${{ steps.meta.outputs.labels }}
          cache-from: type=gha
          cache-to: type=gha,mode=max
          no-cache: ${{ !inputs.use_cache || startsWith(github.ref, 'refs/tags/') }}

      - name: Attest Build Provenance
        if: github.event_name != 'pull_request'
        uses: actions/attest-build-provenance@v1.1.1
        with:
          subject-name: ${{ env.REGISTRY }}/${{ env.IMAGE_NAME }}
          subject-digest: ${{ steps.build-and-push.outputs.digest }}
          push-to-registry: true
```

But the attestation step is failing ([full workflow logs](https://github.com/AFCMS/minetest-skin-server/actions/runs/9051166171/job/24867371956)):

```
Run actions/attest-build-provenance@v1.1.1
Run actions/attest-build-provenance/predicate@db1dde0f270afe12073070ac7aa802958ae3ec04
  
Run actions/attest@32f49af6653ef9b7d1a40182d53fc9ebd53447d8
  
Attestation created for ghcr.io/AFCMS/minetest-skin-server@sha256:36d352f719ccc088efcecf585bb8c93f1588efd896e93d68e78f3f3bfa0751ce
Attestation signed using certificate from Public Good Sigstore instance
Attestation signature uploaded to Rekor transparency log
https://search.sigstore.dev?logIndex=93072090
Attestation uploaded to repository
https://github.com/AFCMS/minetest-skin-server/attestations/82[95](https://github.com/AFCMS/minetest-skin-server/actions/runs/9051166171/job/24867371956#step:8:98)14
Error: Error: Invalid image name: ghcr.io/AFCMS/minetest-skin-server
```

I don't know why, because I followed exactly the steps in the docs, and if the image name was invalid it should already fail in the build step?

**Comment by bdehamer** [comment]
@AFCMS I believe the issue is that image names are not allowed to have uppercase characters in them. I suspect that `docker/metadata-action` and `docker/buid-push-action` are automatically downcasing the supplied name to ensure that the name is valid.

We should probably do something similar . . . but it gets a bit tricky cause we wouldn't want to universally downcase any supplied `subject-name` (case is probably relevant for a good number of subjects). Perhaps we could use the `push-to-registry` flag as a signal that the subject name should be downcased. We'll give it some thought and work on a fix.

In the meantime, you can work around the issue by updating your input value to be downcased.

**Comment by AFCMS** [comment]
I have downcased the value passed to the action and it no longer fails.

But I don't understand something, the action successfully create the attestation in the actions tab and I can use the GH CLI to verify the image with the master tag; but it seems a tag is pushed each time with the SHA of the master branch?

With a push to `master`:

![image](https://github.com/actions/attest-build-provenance/assets/61794590/e85f42d6-d74b-42e1-b0cc-62ebc2fe8792)

It's pretty annoying and probably not wanted to have tagged versions for each push, I tried deleting the tagged SHA version and the verification still works with GH CLI.

**Comment by AFCMS** [comment]
> @AFCMS The `gh` CLI will always use the GitHub API to retrieve the attestation for verification -- this is why verification continues to work even if you remove the attestation from the OCI registry. In the future, we may introduce an option for the CLI to read the attestation from the registry, but this is not supported at the moment.
> 
> As for the extra tag you see, this is a by-product of the way attestations are associated with the image they are attesting. This scheme is described as part of the [OCI specification](https://github.com/opencontainers/distribution-spec/blob/main/spec.md#unavailable-referrers-api) and allows clients to create/discover relationships between container images and arbitrary artifacts which refer to those images.
> 
> Some registries have support for a feature called the "referrers API" which simplifies the process of discovering these referring artifacts, but the "referrers tag schema" (which is what you're seeing here) is used as a fallback when that API is not supported.

So it's basically a fallback ? Doesn't ghcr.io support this API?

Having a tag pushed on each commit is pretty annoying, since the "Install from the command line" from the package page no longer show the real latest tag, but the fallback tag since the fallback tag was created after the other one.

**Comment by bdehamer** [comment]
No, GHCR does not yet support the OCI referrers API.

Unless you have a specific use case for reading the provenance attestation from the registry, my recommendation would be that you omit the `push-to-registry` option.

**Comment by AFCMS** [comment]
Okay, thanks for your time!

I hope ghcr.io will get this API soon :slightly_smiling_face:

**Comment by bdehamer** [comment]
Just published [v1.1.2](https://github.com/actions/attest-build-provenance/releases/tag/v1.1.2) of the action which will automatically downcase the repository name when using the `push-to-registry` option.
